db/mongodb.go

```diff
@@ -0,0 +1,283 @@
+	}
+	if len(empresas) > 0 {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		_, err := collection.InsertMany(ctx, empresas)
+		if err != nil {
+			return fmt.Errorf("error inserting companies into MongoDB: %w", err)
+		}
+	} else {
+		fmt.Println("No valid company to insert.")
+	}
+	return nil
+}
+
+func (m *MongoDB) MetaSave(k, v string) error {
+	if m == nil {
+		return fmt.Errorf("MongoDB connection not initialized")
+	}
+	collection := m.db.Collection(metaTableName)
+	if len(k) > 16 {
+		return fmt.Errorf("the key can have a maximum of 16 characters")
+	}
+	doc := bson.M{
+		"key":   k,
+		"value": v,
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	_, err := collection.InsertOne(ctx, doc)
+	if err != nil {
+		return fmt.Errorf("error saving %s to the meta collection: %w", k, err)
+	}
+	return nil
+}
+
+// MetaRead reads a key/value pair from the metadata collection.
+func (m *MongoDB) MetaRead(k string) (string, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	var result struct {
+		Value string `bson:"value"`
+	}
+	collection := m.db.Collection(metaTableName)
+	err := collection.FindOne(ctx, bson.M{"key": k}).Decode(&result)
+	if err != nil {
+		if err == mongo.ErrNoDocuments {
+			return "", fmt.Errorf("metadata key %s not found", k)
+		}
+		return "", fmt.Errorf("error looking for metadata key %s: %w", k, err)
+	}
+	return result.Value, nil
+}
+
+// Close terminates the connection to MongoDB.
+func (m *MongoDB) Close() error {
+	if err := m.client.Disconnect(m.ctx); err != nil {
+		return fmt.Errorf("error disconnecting from MongoDB: %w", err)
+	}
+	fmt.Println("Successfully disconnected from MongoDB")
+	return nil
+}
+
+// PreLoad runs before starting to load data into the database.
+func (m *MongoDB) PreLoad() error {
+	return nil
+}
+
+// PostLoad runs after loading data into the database.
+// Removes duplicates and creates indexes.
+func (m *MongoDB) PostLoad() error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	collection := m.db.Collection(companyTableName)
+	pipeline := mongo.Pipeline{
+		{{"$group", bson.D{
+			{"_id", "$cnpj"},
+			{"docs", bson.D{{"$push", "$_id"}}},
+			{"count", bson.D{{"$sum", 1}}},
+		}}},
+		{{"$match", bson.D{{"count", bson.D{{"$gt", 1}}}}}},
+	}
+	cursor, err := collection.Aggregate(ctx, pipeline)
+	if err != nil {
+		return fmt.Errorf("error executing aggregation: %w", err)
+	}
+	defer cursor.Close(ctx)
+	// Iterates through the results and removes duplicates
+	for cursor.Next(ctx) {
+		var result struct {
+			ID   string               `bson:"_id"`
+			Docs []primitive.ObjectID `bson:"docs"`
+		}
+		if err := cursor.Decode(&result); err != nil {
+			return fmt.Errorf("error decoding result: %w", err)
+		}
+		// Keep the first document and remove the others
+		if len(result.Docs) > 1 {
+			toRemove := result.Docs[1:] // Delete the first document
```

Comment:
Esses comentários estão contraditórios. Creio que a iseia seja:

```suggestion
		// Keep the first document and remove the others
		if len(result.Docs) > 1 {
			toRemove := result.Docs[1:] // Delete all but the first document
```

Mas aí o segundo comentário diz a mesma coisa que o primeiro. Se esse for o caso:

```suggestion
		// Keep the first document and remove the others
		if len(result.Docs) > 1 {
			toRemove := result.Docs[1:]
```

E, nomes curtos ; )

```suggestion
		// Keep the first document and remove the others
		if len(result.Docs) > 1 {
			ds := result.Docs[1:]
```